troubleshooting/tools/log-loss-test-framework/validator/validate.go

```diff
@@ -0,0 +1,272 @@
+		fmt.Println("[CW Log Loss Validator] Unable to get ECS Metadata, ignore this warning if not running on ECS")
+		return metadata
+	}
+
+	res, err := httpClient.Get(ecsTaskMetadataEndpointV4 + "/task")
+	if err != nil {
+		fmt.Printf("[CW Log Loss Validator] Failed to get ECS Metadata via HTTP Get: %s\n", err)
+		os.Exit(1)
+	}
+
+	response, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		fmt.Printf("[CW Log Loss Validator] Failed to read ECS Metadata from HTTP response: %s\n", err)
+		os.Exit(1)
+	}
+	res.Body.Close()
+
+	err = json.Unmarshal(response, &metadata)
+	if err != nil {
+		fmt.Printf("[CW Log Loss Validator] Failed to unmarshal ECS metadata: %s\n", err)
+		os.Exit(1)
+	}
+
+	arn, err := arn.Parse(metadata.ECS_TASK_ARN)
+	if err != nil {
+		fmt.Printf("[CW Log Loss Validator] Failed to parse ECS TaskARN: %s\n", err)
+		os.Exit(1)
+	}
+
+	resourceID := strings.Split(arn.Resource, "/")
+	taskID := resourceID[len(resourceID)-1]
+	metadata.ECS_TASK_ID = taskID
+	metadata.AWS_REGION = arn.Region
+	metadata.ECS_TASK_DEFINITION = metadata.ECS_FAMILY + ":" + metadata.ECS_REVISION
+
+	return metadata
+}
+
+
+// Creates a new CloudWatch Client
+func getCWClient(region string) (*cloudwatchlogs.CloudWatchLogs, error) {
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(region)},
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return cloudwatchlogs.New(sess), nil
+}
+
+// Validate logs in CloudWatch.
+// Similar logic as S3 validation.
+func validateCloudwatch(cwClient *cloudwatchlogs.CloudWatchLogs, logGroup string, logStream string, inputMap map[string]bool) (int, map[string]bool) {
+	var forwardToken *string
+	var input *cloudwatchlogs.GetLogEventsInput
+	cwRecoredCounter := 0
+
+	// Returns all log events from a CloudWatch log group with the given log stream.
+	// This approach utilizes NextForwardToken to pull all log events from the CloudWatch log group.
+	for {
+		if forwardToken == nil {
+			input = &cloudwatchlogs.GetLogEventsInput{
+				LogGroupName:  aws.String(logGroup),
+				LogStreamName: aws.String(logStream),
+				StartFromHead: aws.Bool(true),
+			}
+		} else {
+			input = &cloudwatchlogs.GetLogEventsInput{
+				LogGroupName:  aws.String(logGroup),
+				LogStreamName: aws.String(logStream),
+				NextToken:     forwardToken,
+				StartFromHead: aws.Bool(true),
+			}
+		}
+
+		response, err := cwClient.GetLogEvents(input)
+		for err != nil {
+			// retry for throttling exception
+			if strings.Contains(err.Error(), "ThrottlingException: Rate exceeded") {
+				time.Sleep(1 * time.Second)
+				response, err = cwClient.GetLogEvents(input)
+			} else {
+				exitErrorf("[TEST FAILURE] Error occured to get the log events from log group: %q., %v", logGroup, err)
+			}
+		}
+
+		for _, event := range response.Events {
+			log := aws.StringValue(event.Message)
+
+			// First 8 char is the unique record ID
+			recordId := log[:8]
+			cwRecoredCounter += 1
+			if _, ok := inputMap[recordId]; ok {
```

Comment:
Just a note: if you run a 10mbps test for 1 day this map would be 6.91200GB if my math is correct:
10000 logs/seconds * 8 bytes/log * 86400 seconds  =  6.912Gb. If you see oom kill it may be because of this map.

Reply:
this is part of why each test run only sends a few hundred GB of data and runs for only a few minutes